File: articles/differential-equations-for-systems-modeling/cpp/logistic_ode.cpp
#include <iostream>
#include <vector>

double logistic_rate(double state, double growth_rate, double capacity) {
    return growth_rate * state * (1.0 - state / capacity);
}

std::vector<double> simulate_logistic(double initial_state, double growth_rate, double capacity, double dt, int steps) {
    std::vector<double> state(steps, 0.0);
    state[0] = initial_state;

    for (int i = 1; i < steps; ++i) {
        double derivative = logistic_rate(state[i - 1], growth_rate, capacity);
        state[i] = state[i - 1] + derivative * dt;
    }

    return state;
}

int main() {
    auto state = simulate_logistic(10.0, 0.20, 100.0, 0.1, 300);

    std::cout << "Final state: " << state.back() << "\n";

    return 0;
}

File: articles/differential-equations-for-systems-modeling/go/logistic_ode.go
package main

import "fmt"

func logisticRate(state, growthRate, capacity float64) float64 {
	return growthRate * state * (1.0 - state/capacity)
}

func simulateLogistic(initialState, growthRate, capacity, dt float64, steps int) []float64 {
	state := make([]float64, steps)
	state[0] = initialState

	for i := 1; i < steps; i++ {
		derivative := logisticRate(state[i-1], growthRate, capacity)
		state[i] = state[i-1] + derivative*dt
	}

	return state
}

func main() {
	state := simulateLogistic(10.0, 0.20, 100.0, 0.1, 300)
	fmt.Printf("Final state: %.6f\n", state[len(state)-1])
}
